### PR TITLE
Add TextBufferProjectService Mac implementation.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Was unable to test the text buffer project service class due to signing issues with the MonoDevelop.Sdk dependency.

#1789

### CODE REVIEW

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                throw new ArgumentNullException(nameof(textBuffer));
+            }
+
+            // If there's no document we can't find the FileName, or look for an associated project.
+            if (!_documentFactory.TryGetTextDocument(textBuffer, out var textDocument))
+            {
+                return null;
+            }
+
+            var projectsContainingFilePath = _workspace.MonoDevelopSolution.GetProjectsContainingFile(textDocument.FilePath);
```
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: Today this always returns 0 projects. Will need to wait on the MonoDevelop folks to fill in the Workspace with the right information.

If you want to test this manually you can replace this line with:
`var projectsContainingFilePath = IdeApp.Workspace.GetProjectsContainingFile(textDocument.FilePath);`

Comment (rynowak) [reply src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: Ah I see. Yeah I wasn't sure which project service you were talking about. This class shouldn't rely on roslyn. In VS (windows) this uses `IVsHierarchy`.

**Comment by NTaylorMullen** [comment]
🆙 📅

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var dotnetProject = project as DotNetProject;
+
+            Debug.Assert(dotnetProject != null);
+
+            return dotnetProject.FileName.FullPath;
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: This should be a c-style cast, it's going to throw anyway

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        public override bool IsSupportedProject(object project)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var dotnetProject = (DotNetProject)project;
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: 👍

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+            Debug.Assert(dotnetProject != null);
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: ehhhh this is kinda ehhhh

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                if (dotnetProject.IsCapabilityMatch(AspNetCoreWebCapability))
+                {
+                    return true;
+                }
+
+                // ASP.NET Core Class libraries
+                if (dotnetProject.References.Any(reference => reference.Reference.StartsWith(MvcAssemblyName, StringComparison.Ordinal)))
+                {
+                    return true;
+                }
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: Do we need this at all? Why not just `return true`. There no non-Core Razor in VS4Mac

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            var dotnetProject = (DotNetProject)project;
+
+            Debug.Assert(dotnetProject != null);
+
+            return dotnetProject.Name;
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: I think you may have been on something when you wrote these asserts.

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs (hunk 4 above)
Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: Yup copy paste flub about this.

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs (hunk 2 above)
Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: sure

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs (hunk 5 above)
Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: We could. Probably being a little bit overprotective here.

File: src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs (hunk 6 above)
Comment (NTaylorMullen) [reply src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/Editor/DefaultTextBufferProjectService.cs]: lol looking back at these totally agree. Good submission to shitcode 😉

**Pull request merged**